### PR TITLE
feat(dataset): dataset provenance in a single file

### LINKED ISSUE #1710: Store all datasets in single file
Similar to what was done in https://github.com/SwissDataScienceCenter/renku-python/pull/1545 we should add dataset metadata for the current commit in a single file. Dataset commands should write to this new file as well as to the old location. Commands still use the old format.

We keep the old files for compatibility with `renku log` but exclude the new file from it.

### PR DESCRIPTION
**Pull request by mohammad-sdsc**
# Description

Parses git history and puts all datasets' metadata in a single file: `.renku/dataset.json`. To generate this file use: `renku graph generate-dataset`. To export add `-d` flag to the export command: `renku graph export -d --format jsonld`.

Fixes #1710

### CODE REVIEW

File: renku/core/models/provenance/datasets.py [hunk 1]
```diff
@@ -0,0 +1,515 @@
...
+
+        rdf_type = schema.Dataset
+        model = Dataset
+        unknown = EXCLUDE
+
+    creators = Nested(schema.creator, PersonSchema, many=True)
+    date_created = fields.DateTime(
+        schema.dateCreated, missing=None, allow_none=True, format="iso", extra_formats=("%Y-%m-%d",),
+    )
+    date_deleted = fields.DateTime(prov.invalidatedAtTime, missing=None, allow_none=True, format="iso")
```
Comment (Panaetius) [inline renku/core/models/provenance/datasets.py:481]: If we want to be consistent with PROV-O, then to use `invalidatedAtTime`, dataset also needs to be a `prov.entity`, as that's the domain of that property.

File: renku/core/models/provenance/datasets.py [hunk 2]
```diff
@@ -0,0 +1,515 @@
...
+            u = urlparse(derived_from_url)
+            derived_from_url = u._replace(netloc=host).geturl()
+            self.derived_from = Url(id=derived_from_id, url_id=derived_from_url)
+
+        for file_ in self.files:
+            file_.client = client
+
+    def update_from(self, dataset, client, revision, date):
+        """Update metadata from a new version of the dataset."""
+        self._set_identifier(dataset.identifier)
```
Comment (Panaetius) [inline renku/core/models/provenance/datasets.py:274]: Does this go against dataset immutability? I.e. using this you can change the properties of a dataset without generating a new ID?

File: renku/core/incubation/graph.py [hunk 3]
```diff
@@ -217,3 +230,120 @@ def _get_modified_paths(client, plans_usages):
...
+    for n, commit in enumerate(commits, start=1):
+        communication.echo(f"\rProcessing commits {n}/{n_commits}\r")
+
+        files_diff = list(commit.diff(commit.parents or NULL_TREE, paths=".renku/datasets/*"))
+        paths = [git_unicode_unescape(f.a_path) for f in files_diff]
+        deleted_paths = [git_unicode_unescape(f.a_path) for f in files_diff if f.change_type == "A"]
+
+        datasets, deleted_datasets = _fetch_datasets(client, commit.hexsha, paths=paths, deleted_paths=deleted_paths)
+
+        revision = commit.hexsha
```
[the remark this answers is not in the document]
Comment (mohammad-sdsc) [reply renku/core/incubation/graph.py:264]: We use git object hashes and not git commit hashes. Revision is passed only to enable calculating the correct content hash in git history at a specific revision. For example, when a file was added to a dataset at revision xyz we want to calculate file's content hash at that revision (and not its current content).

File: renku/core/models/provenance/datasets.py (hunk 2 above)
Comment (mohammad-sdsc) [reply renku/core/models/provenance/datasets.py:274]: Right. Actually, when dataset has a non-existing identifier we insert it as a new dataset and this function won't get called. This is called only when identifier is the same as the current identifier, so, we don't need to set it again.

File: renku/core/incubation/graph.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (Panaetius) [reply renku/core/incubation/graph.py:265]: Ah I see :+1:

**Review by Panaetius — DISMISSED** [review]
Thank you, looks great!

**Review by Panaetius — APPROVED** [review]

**Pull request merged**